File: apps/mobile/lib/core/user/models/app_capability.dart
enum CapabilitySubject {
  calendar,
  catalog,
  cadastroSubmission,
  fieldSuggestion,
  facility,
  person,
  territory,
  user,
  visit,
}

enum CapabilityAction { read, create, update, delete, manage }

extension CapabilitySubjectX on CapabilitySubject {
  String get wireName => switch (this) {
    .calendar => 'CALENDAR',
    .catalog => 'CATALOG',
    .cadastroSubmission => 'CADASTRO_SUBMISSION',
    .fieldSuggestion => 'FIELD_SUGGESTION',
    .facility => 'FACILITY',
    .person => 'PERSON',
    .territory => 'TERRITORY',
    .user => 'USER',
    .visit => 'VISIT',
  };

  static CapabilitySubject? tryParse(String value) {
    for (final subject in CapabilitySubject.values) {
      if (subject.wireName == value) return subject;
    }
    return null;
  }
}

extension CapabilityActionX on CapabilityAction {
  String get wireName => name;

  static CapabilityAction? tryParse(String value) {
    for (final action in CapabilityAction.values) {
      if (action.wireName == value) return action;
    }
    return null;
  }
}

class CapabilityRule {
  const CapabilityRule({
    required this.action,
    required this.subject,
    this.inverted = false,
  });

  final CapabilityAction action;
  final CapabilitySubject subject;
  final bool inverted;
}

File: packages/access/src/capabilities/app-capability.ts
import type { AccessGrantRecord } from "../contracts/access-grant.contract";
import type { Role } from "../enums/role.enum";
import { defineAbilitiesForUser } from "../permissions/grant.permissions";

export const APP_CAPABILITIES = [
  "agenda.read",
  "agenda.create",
  "agenda.update",
  "agenda.delete",
  "catalog.read",
  "catalog.manage",
  "cadastro.read",
  "cadastro.review",
  "field-suggestion.read",
  "field-suggestion.review",
  "facility.read",
  "facility.create",
  "facility.update",
  "facility.delete",
  "professional.read",
  "professional.update",
  "territory.read",
  "territory.create",
  "territory.update",
  "territory.delete",
  "user.read",
  "user.manage",
  "user.lifecycle",
] as const;

export type AppCapability = (typeof APP_CAPABILITIES)[number];

const CAPABILITY_CHECKS: Record<AppCapability, { action: string; subject: string }> = {
  "agenda.read": { action: "read", subject: "CALENDAR" },
  "agenda.create": { action: "create", subject: "CALENDAR" },
  "agenda.update": { action: "update", subject: "CALENDAR" },
  "agenda.delete": { action: "delete", subject: "CALENDAR" },
  "catalog.read": { action: "read", subject: "CATALOG" },
  "catalog.manage": { action: "manage", subject: "CATALOG" },
  "cadastro.read": { action: "read", subject: "CADASTRO_SUBMISSION" },
  "cadastro.review": { action: "update", subject: "CADASTRO_SUBMISSION" },
  "field-suggestion.read": { action: "read", subject: "FIELD_SUGGESTION" },
  "field-suggestion.review": { action: "update", subject: "FIELD_SUGGESTION" },
  "facility.read": { action: "read", subject: "FACILITY" },
  "facility.create": { action: "create", subject: "FACILITY" },
  "facility.update": { action: "update", subject: "FACILITY" },
  "facility.delete": { action: "delete", subject: "FACILITY" },
  "professional.read": { action: "read", subject: "PROFESSIONAL" },
  "professional.update": { action: "update", subject: "PROFESSIONAL" },
  "territory.read": { action: "read", subject: "TERRITORY" },
  "territory.create": { action: "create", subject: "TERRITORY" },
  "territory.update": { action: "update", subject: "TERRITORY" },
  "territory.delete": { action: "delete", subject: "TERRITORY" },
  "user.read": { action: "read", subject: "USER" },
  "user.manage": { action: "manage", subject: "USER" },
  "user.lifecycle": { action: "update", subject: "USER" },
};

export function getAppCapabilities(role: Role, grants: AccessGrantRecord[] = []): AppCapability[] {
  const ability = defineAbilitiesForUser(role, grants);
  return APP_CAPABILITIES.filter((capability) => {
    const check = CAPABILITY_CHECKS[capability];
    return ability.can(check.action as any, check.subject as any);
  });
}
